core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java

```diff
@@ -252,7 +252,7 @@ public synchronized void schedulingComplete()
             return;
         }
 
-        if (getAllTasks().stream().anyMatch(task -> getState() == StageState.RUNNING)) {
+        if (getAllTasks().stream().anyMatch(task -> task.getTaskStatus().getState() == TaskState.RUNNING)) {
```

Comment:
> @sopel39 the code before the change clearly uses a wrong condition, since was calling `this.getState()` for each task.
> 
> @guhanjie the transition seems to happen here
> https://github.com/trinodb/trino/blob/77a24a07beed1752014d28ec7cc4294ab4875d36/core/trino-main/src/main/java/io/trino/execution/SqlStageExecution.java#L545
> 
> maybe instead of fixing this line, we can remove it? (thinking loud)
> 
> @dain ptal

yes, there are two points to RUNNING, one is we talking about, the other one is you mentioned that, which can make sure the Stage state transition to RUNNING. Actually, all achieve to RUNNING is go through this way

Reply:
@guhanjie i understand you found the bug by looking at the code -- via code inspection. 
the code should be fixed, the question is -- what should be the fix.

if we're changing the code to make the condition be applicable, we should ask ourselves, what problem are we solving from end-user perspective. ie. is there a functional bug?
I don't feel safe proceeding without answer to that question, and i don't know the answer.

If we remove the code as dead code  -- as is currently is -- we don't make a behavioral difference while we also improve readability, and that's something i would approve right away.

Reply:
agree with you, i removed the code(never worked currently), it is safe and can reduce the tasks iteration checks and code is clear(there is only one point to RUNNING)
and as i mentioned above, i also think the state change latency would be no impact, 
because the first point faster than the second point obviously is such rare in fact.

I'm pleasure to see more looks or talks before the change code is approved. @dain  @sopel39